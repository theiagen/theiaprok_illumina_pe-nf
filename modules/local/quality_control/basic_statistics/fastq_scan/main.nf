process FASTQ_SCAN_PE {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/biocontainers/fastq-scan:1.0.1--h4ac6f70_3"

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path("*_R1_fastq-scan.json"), emit: read1_fastq_scan_json
    tuple val(meta), path("*_R2_fastq-scan.json"), emit: read2_fastq_scan_json
    tuple val(meta), path("READ1_SEQS"), emit: read1_seq
    tuple val(meta), path("READ2_SEQS"), emit: read2_seq
    tuple val(meta), path("READ_PAIRS"), emit: read_pairs
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def read1 = reads[0]
    def read2 = reads[1]
    if (!read1 || !read2) {
        error "FASTQ_SCAN_PE requires two input reads, but only found: ${reads}"
    }
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    """
    # exit task in case anything fails in one-liners or variables are unset
    set -euo pipefail
    
    # set cat command based on compression
    if [[ "${read1}" == *".gz" ]] ; then
        cat_reads="zcat"
    else
        cat_reads="cat"
    fi
    
    # capture forward read stats
    echo "DEBUG: running fastq-scan on \$(basename ${read1})"
    \${cat_reads} ${read1} | fastq-scan | tee ${prefix}_R1_fastq-scan.json
    
    # using jq to extract read count
    read1_seqs=\$(jq .qc_stats.read_total ${prefix}_R1_fastq-scan.json)
    echo "DEBUG: number of reads in \$(basename ${read1}): \${read1_seqs}"
    echo
    
    # capture reverse read stats
    echo "DEBUG: running fastq-scan on \$(basename ${read2})"
    \${cat_reads} ${read2} | fastq-scan | tee ${prefix}_R2_fastq-scan.json
    
    # using jq to extract read count
    read2_seqs=\$(jq .qc_stats.read_total ${prefix}_R2_fastq-scan.json)
    echo "DEBUG: number of reads in \$(basename ${read2}): \${read2_seqs}"
    
    # capture number of read pairs
    if [ "\${read1_seqs}" == "\${read2_seqs}" ]; then
        read_pairs="\${read1_seqs}"
    else
        read_pairs="Uneven pairs: R1=\${read1_seqs}, R2=\${read2_seqs}"
    fi
    
    echo "DEBUG: number of read pairs: \${read_pairs}"
    
    # Set environment variables for outputs
    READ1_SEQS="\$read1_seqs"
    READ2_SEQS="\$read2_seqs"
    READ_PAIRS="\$read_pairs"

    # Write read counts to files
    echo "\$READ1_SEQS" > READ1_SEQS
    echo "\$READ2_SEQS" > READ2_SEQS
    echo "\$READ_PAIRS" > READ_PAIRS

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastq-scan: \$(fastq-scan -v 2>&1 | sed 's/fastq-scan //')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo '{"qc_stats":{"read_total":1000}}' > ${prefix}_R1_fastq-scan.json
    echo '{"qc_stats":{"read_total":1000}}' > ${prefix}_R2_fastq-scan.json
    
    READ1_SEQS="1000"
    READ2_SEQS="1000"
    READ_PAIRS="1000"

    echo "\$READ1_SEQS" > READ1_SEQS
    echo "\$READ2_SEQS" > READ2_SEQS
    echo "\$READ_PAIRS" > READ_PAIRS

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastq-scan: 1.0.1
    END_VERSIONS
    """
}