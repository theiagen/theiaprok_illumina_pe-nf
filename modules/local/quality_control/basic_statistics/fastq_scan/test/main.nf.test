nextflow_process {
    name "Test Process FASTQ_SCAN_PE"
    script "../main.nf"
    process "FASTQ_SCAN_PE"
    
    test("fastq_scan pe test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.read1_fastq_scan_json,
                    process.out.read2_fastq_scan_json,
                    process.out.read1_seq,
                    process.out.read2_seq,
                    process.out.read_pairs,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("should run with theiaprok workflow - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}