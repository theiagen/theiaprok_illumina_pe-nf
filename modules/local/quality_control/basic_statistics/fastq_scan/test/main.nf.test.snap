{
    "fastq_scan pe test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastq-scan.json:md5,71dc2588c0cc06ea03189775de2e9755"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastq-scan.json:md5,5a61069d962fe7c6cf961a4fc455f77e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "READ1_SEQS:md5,2ebd3caa45b31a2e74e436b645eab4b0"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "READ2_SEQS:md5,2ebd3caa45b31a2e74e436b645eab4b0"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "READ_PAIRS:md5,2ebd3caa45b31a2e74e436b645eab4b0"
                ]
            ],
            [
                "versions.yml:md5,f43574e4a33b4589e98554dae663cfd0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T18:31:53.009538888"
    },
    "should run with theiaprok workflow - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R1_fastq-scan.json:md5,98de982178be794f9db4ea493683a5da"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R2_fastq-scan.json:md5,98de982178be794f9db4ea493683a5da"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ1_SEQS:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ2_SEQS:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ_PAIRS:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "5": [
                    "versions.yml:md5,f43574e4a33b4589e98554dae663cfd0"
                ],
                "read1_fastq_scan_json": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R1_fastq-scan.json:md5,98de982178be794f9db4ea493683a5da"
                    ]
                ],
                "read1_seq": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ1_SEQS:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "read2_fastq_scan_json": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R2_fastq-scan.json:md5,98de982178be794f9db4ea493683a5da"
                    ]
                ],
                "read2_seq": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ2_SEQS:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "read_pairs": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ_PAIRS:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f43574e4a33b4589e98554dae663cfd0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T18:32:03.936097584"
    }
}